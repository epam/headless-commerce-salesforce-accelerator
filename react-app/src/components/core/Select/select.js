/*
 * Copyright 2020 EPAM Systems, Inc. (https://www.epam.com/)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState } from "react";
import PropTypes from "prop-types";
import StyledTypography from "components/core/Typography";
import {
  StyledFormControl,
  StyledSelect,
  StyledSelectWrapper,
  StyledErrorMsg,
} from "./selectStyled";
import ArrowIcon from "../ArrowIcon";

const Select = ({
  IconComponent = ArrowIcon,
  defaultValue = "",
  handleChange = () => {},
  values,
  firstOptionContent,
  width,
  height,
  label,
  iconisrotated,
  error = false,
  errorMsg,
  fullWidth,
  margin,
  name = "",
  register,
  required,
}) => {
  const [state, setState] = useState(defaultValue || "");

  return (
    <StyledSelectWrapper margin={margin}>
      <StyledTypography margin="0 0 5px 0">{label}</StyledTypography>
      <StyledFormControl
        variant="outlined"
        margin="dense"
        width={width}
        height={height}
        focused={false}
        error={error}
        fullWidth={fullWidth}
        required={required}
      >
        <StyledSelect
          native
          IconComponent={IconComponent}
          value={state}
          iconisrotated={iconisrotated?.toString()}
          onChange={(e) => {
            setState(e.target.value);
            handleChange(e);
          }}
          inputRef={
            register &&
            register({
              required,
            })
          }
          name={name}
        >
          {firstOptionContent && <option value="">{firstOptionContent}</option>}
          {values.map((item) => (
            <option
              key={item.id || item.ID || item.value || item.label}
              value={item.value || item.ID || ""}
              disabled={"selectable" in item ? !item.selectable : false}
            >
              {item.label || item.displayValue || item.value}
            </option>
          ))}
        </StyledSelect>
      </StyledFormControl>
      {error && errorMsg ? <StyledErrorMsg>{errorMsg}</StyledErrorMsg> : null}
    </StyledSelectWrapper>
  );
};

Select.propTypes = {
  IconComponent: PropTypes.elementType,
  defaultValue: PropTypes.string,
  handleChange: PropTypes.func,
  values: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      selectable: PropTypes.bool,
      displayValue: PropTypes.string,
    })
  ),
  firstOptionContent: PropTypes.string,
  width: PropTypes.number,
  height: PropTypes.number,
  label: PropTypes.string,
  iconisrotated: PropTypes.bool,
  fullWidth: PropTypes.bool,
  margin: PropTypes.string,
  error: PropTypes.bool,
  errorMsg: PropTypes.string,
  name: PropTypes.string,
  register: PropTypes.func,
  required: PropTypes.bool,
};

export default Select;
