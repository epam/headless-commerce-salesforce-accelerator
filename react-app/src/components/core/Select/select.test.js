/*
 * Copyright 2020 EPAM Systems, Inc. (https://www.epam.com/)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react";
import { shallow } from "enzyme";
import Select from ".";
import {
  StyledFormControl,
  StyledSelect,
  StyledSelectWrapper,
  StyledErrorMsg,
} from "./selectStyled";
import getMainTheme from "../../../theme";

const defaultProps = {
  values: [
    {
      ID: "001",
      default: true,
      description: "Order received within 7-10 business days",
      displayName: "Ground",
      displayValue: "Ground (7-10 Business Days)",
      estimatedArrivalTime: "7-10 Business Days",
      selected: true,
      shippingCost: "$7.99",
    },
    {
      ID: "002",
      default: false,
      description: "Order received in 2 business days",
      displayName: "2-Day Express",
      displayValue: "2-Day Express (2 Business Days)",
      estimatedArrivalTime: "2 Business Days",
      selected: false,
      shippingCost: "$11.99",
    },
  ],
  iconisrotated: false,
  margin: "16px 0px 0px 0px",
  fullWidth: true,
};

const theme = getMainTheme();

describe("ArrowIcon", () => {
  const wrapped = shallow(<Select {...defaultProps} />);
  it("should match snapshot", () => {
    expect(wrapped).toMatchSnapshot();
  });
});

describe("StyledSelect", () => {
  const wrapped = shallow(<StyledSelect {...defaultProps} theme={theme} />);
  it("should match snapshot", () => {
    expect(wrapped).toMatchSnapshot();
  });
});

describe("StyledFormControl", () => {
  const wrapped = shallow(<StyledFormControl theme={theme} />);
  it("should match snapshot", () => {
    expect(wrapped).toMatchSnapshot();
  });
});

describe("StyledErrorMsg", () => {
  const wrapped = shallow(<StyledErrorMsg theme={theme} />);
  it("should match snapshot", () => {
    expect(wrapped).toMatchSnapshot();
  });
});

describe("StyledSelectWrapper", () => {
  const wrapped = shallow(
    <StyledSelectWrapper theme={theme} {...defaultProps} />
  );
  it("should match snapshot", () => {
    expect(wrapped).toMatchSnapshot();
  });
});
