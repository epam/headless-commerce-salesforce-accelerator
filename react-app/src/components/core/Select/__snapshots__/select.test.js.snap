// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrowIcon should match snapshot 1`] = `
<styled.div
  margin="16px 0px 0px 0px"
>
  <Styled(WithStyles(ForwardRef(Typography)))
    margin="0 0 5px 0"
  />
  <Styled(WithStyles(ForwardRef(FormControl)))
    error={false}
    focused={false}
    fullWidth={true}
    margin="dense"
    variant="outlined"
  >
    <Styled(WithStyles(ForwardRef(Select)))
      IconComponent={[Function]}
      iconisrotated="false"
      name=""
      native={true}
      onChange={[Function]}
      value=""
    >
      <option
        disabled={false}
        key="001"
        value="001"
      >
        Ground (7-10 Business Days)
      </option>
      <option
        disabled={false}
        key="002"
        value="002"
      >
        2-Day Express (2 Business Days)
      </option>
    </Styled(WithStyles(ForwardRef(Select)))>
  </Styled(WithStyles(ForwardRef(FormControl)))>
</styled.div>
`;

exports[`StyledErrorMsg should match snapshot 1`] = `
<div
  className="sc-AxhUy gkiEFc"
/>
`;

exports[`StyledFormControl should match snapshot 1`] = `
<WithStyles(ForwardRef(FormControl))
  className="sc-AxirZ OLvNS"
  theme={
    Object {
      "breadcrumbsH": 40,
      "breakpoints": Object {
        "mobile": 992,
      },
      "colLg": 74,
      "colMd": 69,
      "colSm": 60,
      "colors": Object {
        "alabaster": "#FBFBFB",
        "amber": "#FFC000",
        "athensGray": "#FAFBFC",
        "black": "#000",
        "blue": "#376A9A",
        "dimGray": "#CBCDCE",
        "gullGrey": "#A7A9AA",
        "iron": "#D9DCDD",
        "lightGray": "#DEE2E3",
        "mineShaft": "#222222",
        "miscellaneous": "linear-gradient(90deg, #E77878 0%, #EFD063 25.00%, #65D5EE 50.00%, #C083E6 75.00%, #7F6EE8 100%);",
        "pale": "#FFEFEF",
        "punch": "#DA3232",
        "regentGrey": "#7D8081",
        "rollingStone": "#7E8182",
        "semiWhite": "rgba(255, 255, 255, 0.75)",
        "silverChalice": "#A9A9A9",
        "turmeric": "#C5D153",
        "wattle": "#CEDB56",
        "whisper": "#EBEBEB",
        "white": "#FFFFFF",
        "yellowGreen": "#D8E278",
      },
      "contentWidth": 1200,
      "font": [Function],
      "fontSize": 14,
      "footerH": 331,
      "gutter": 24,
      "headerH": 160,
      "navBarH": 56,
      "pxToRem": [Function],
      "searchBarH": 72,
      "topBarH": 32,
    }
  }
/>
`;

exports[`StyledSelect should match snapshot 1`] = `
<WithStyles(ForwardRef(Select))
  className="sc-AxiKw fgWQuF"
  fullWidth={true}
  iconisrotated={false}
  margin="16px 0px 0px 0px"
  theme={
    Object {
      "breadcrumbsH": 40,
      "breakpoints": Object {
        "mobile": 992,
      },
      "colLg": 74,
      "colMd": 69,
      "colSm": 60,
      "colors": Object {
        "alabaster": "#FBFBFB",
        "amber": "#FFC000",
        "athensGray": "#FAFBFC",
        "black": "#000",
        "blue": "#376A9A",
        "dimGray": "#CBCDCE",
        "gullGrey": "#A7A9AA",
        "iron": "#D9DCDD",
        "lightGray": "#DEE2E3",
        "mineShaft": "#222222",
        "miscellaneous": "linear-gradient(90deg, #E77878 0%, #EFD063 25.00%, #65D5EE 50.00%, #C083E6 75.00%, #7F6EE8 100%);",
        "pale": "#FFEFEF",
        "punch": "#DA3232",
        "regentGrey": "#7D8081",
        "rollingStone": "#7E8182",
        "semiWhite": "rgba(255, 255, 255, 0.75)",
        "silverChalice": "#A9A9A9",
        "turmeric": "#C5D153",
        "wattle": "#CEDB56",
        "whisper": "#EBEBEB",
        "white": "#FFFFFF",
        "yellowGreen": "#D8E278",
      },
      "contentWidth": 1200,
      "font": [Function],
      "fontSize": 14,
      "footerH": 331,
      "gutter": 24,
      "headerH": 160,
      "navBarH": 56,
      "pxToRem": [Function],
      "searchBarH": 72,
      "topBarH": 32,
    }
  }
  values={
    Array [
      Object {
        "ID": "001",
        "default": true,
        "description": "Order received within 7-10 business days",
        "displayName": "Ground",
        "displayValue": "Ground (7-10 Business Days)",
        "estimatedArrivalTime": "7-10 Business Days",
        "selected": true,
        "shippingCost": "$7.99",
      },
      Object {
        "ID": "002",
        "default": false,
        "description": "Order received in 2 business days",
        "displayName": "2-Day Express",
        "displayValue": "2-Day Express (2 Business Days)",
        "estimatedArrivalTime": "2 Business Days",
        "selected": false,
        "shippingCost": "$11.99",
      },
    ]
  }
/>
`;

exports[`StyledSelectWrapper should match snapshot 1`] = `
<div
  className="sc-AxhCb ebpXXp"
  values={
    Array [
      Object {
        "ID": "001",
        "default": true,
        "description": "Order received within 7-10 business days",
        "displayName": "Ground",
        "displayValue": "Ground (7-10 Business Days)",
        "estimatedArrivalTime": "7-10 Business Days",
        "selected": true,
        "shippingCost": "$7.99",
      },
      Object {
        "ID": "002",
        "default": false,
        "description": "Order received in 2 business days",
        "displayName": "2-Day Express",
        "displayValue": "2-Day Express (2 Business Days)",
        "estimatedArrivalTime": "2 Business Days",
        "selected": false,
        "shippingCost": "$11.99",
      },
    ]
  }
/>
`;
